### PR TITLE
x86: change the APUx leds gpio handling

### PR DESCRIPTION
**Pull request by feckert**
I have made the following changes.

- Removed the package leds-apu2 from **package/kernel/**. This is replaced by the new upstream **leds-apu** driver since kernel version 4.15.
- Add a patchset to backport-4.14 for the leds-apu driver. This was added to 4.15.
- Add a patch to pending-4.14 for the new gpio-apu driver. Should go upstream to the next kernel 4.19.
Must be reviewed first. (https://lkml.org/lkml/2018/8/1/375)
- Set the new config option CONFIG_LEDS_APU and CONFIG_GPIO_APU default to 'not set' on x86 target
- Add a package description for leds-apu to **package/kernel/linux/module/leds.mk**
- Add a package description for gpio-apu to **package/kernel/linux/module/other.mk**

### CODE REVIEW

**Comment by miczyg1** [comment]
Along with @pietrushnic, we are firmware developers from 3mdeb company responsible for PC Engines firmware development and platform maintaining. 

@feckert I also suggest to add support for apu4 boards. They have the same GPIO layout as apu3. Have a look here (page 7):
https://www.pcengines.ch/schema/apu4b.pdf
https://www.pcengines.ch/schema/apu3c.pdf

Regarding the DMI matching. The situation is quite complicated since we had wrong SMBIOS naming scheme for some time and we have just fixed it few months ago. Your code still contains very old naming containing capital letters. We do not release such binaries.
Currently DMI product names (depending on BIOS version) found on PC Engines platforms are:

-PC Engines apu1
-apu1
-PC Engines apu2
-apu2
-PC Engines apu3
-apu3
-PC Engines apu4
-apu4

@feckert the ones with shorter names (apu1, apu2 etc.) are official SMBIOS entries that are correct and found on newest firmwares. They won't ever change in the future. The ones with capital letters: APU, APU2, APU3 IMHO should be removed. @pietrushnic please share Your opinion.

**Comment by miczyg1** [comment]
@feckert I also find that the reset button does not work. I attach `lsmod`: https://justpaste.it/45bhg

**Comment by miczyg1** [comment]
@feckert if the button definition on alix boards conforms to kernel code style, I'm fine with moving the definition of apu button to gpio_apu module.

File: target/linux/generic/pending-4.14/930-gpio-add-support-for-apu2-apu3-devices.patch [hunk 1]
```diff
@@ -0,0 +1,395 @@
+--- /dev/null
```
Comment (blogic) [inline target/linux/generic/pending-4.14/930-gpio-add-support-for-apu2-apu3-devices.patch]: why is this inside generic and not x86 ?

File: package/kernel/linux/modules/leds.mk [hunk 2]
```diff
@@ -145,3 +145,19 @@ define KernelPackage/leds-pca963x/description
 endef
```
Comment (blogic) [inline package/kernel/linux/modules/leds.mk]: this shuould be added to target/linux/x86/modules.mk

File: package/kernel/linux/modules/leds.mk [hunk 3]
```diff
@@ -145,3 +145,19 @@ define KernelPackage/leds-pca963x/description
...
 $(eval $(call KernelPackage,leds-pca963x))
+
+
+define KernelPackage/leds-apu
+  SUBMENU:=$(LEDS_MENU)
+  TITLE:=PC Engines APU/APU2/APU3 LED support
+  DEPENDS:=@TARGET_x86
+  KCONFIG:=CONFIG_LEDS_APU
+  FILES:=$(LINUX_DIR)/drivers/leds/leds-apu.ko
+  AUTOLOAD:=$(call AutoLoad,60,leds-apu,1)
```
Comment (blogic) [inline package/kernel/linux/modules/leds.mk]: cant you autoprobe this one in preinit ?

File: target/linux/generic/pending-4.14/930-gpio-add-support-for-apu2-apu3-devices.patch (hunk 1 above)
Comment (feckert) [reply target/linux/generic/pending-4.14/930-gpio-add-support-for-apu2-apu3-devices.patch]: Yes you are right this must go to
> target/linux/x86/patches-4.14/

File: package/kernel/linux/modules/leds.mk (hunk 2 above)
Comment (feckert) [reply package/kernel/linux/modules/leds.mk]: Ok will do this

File: package/kernel/linux/modules/leds.mk (hunk 3 above)
Comment (feckert) [reply package/kernel/linux/modules/leds.mk]: Still have todo this

**Comment by feckert** [comment]
@pietrushnic as explained in the PR i am working also on this driver to get them merged upstream in to Linux. After the first send i update my driver with the hints from the LKML. But on the mailing-list a question raised up why not load the "APU_FCH_ACPI_MMIO_BASE" address from teh ACPI subsystem !?
I have no idear what they want! I have copied this definitions from the leds-apu.c driver which is already applied upstream. As a coreboot developer for the PC-Engine apu platform could you to solve this or explain why we need to define this in the source.

Discussion on the LKML:
https://lkml.org/lkml/2018/8/1/375

Thanks for reply

**Comment by pietrushnic** [comment]
@feckert let's try to handle that directly on mailing list, if ACPI is broken I'm fine with fixing it according to Linux kernel needs. Let's first start with requested ACPI tables dump - let me handle that.

**Comment by pietrushnic** [comment]
@feckert things take time in mainline. I think Andy should reply if ACPI tables are fine in his opinion. I can fix ACPI if we can clarify requirements.

**Comment by feckert** [comment]
@pietrushnic i am not sure if something happens there in the future. We should ping Andy again?

**Comment by pietrushnic** [comment]
@feckert you can ping him. Maybe someone else can drive what we should do - I'm willing to help enabling that support in mainline, but maintainers should tell us how to attack the problem. Did you added all people form `get_maintainers` script to CC?

**Comment by feckert** [comment]
@blogic I have update my PR with the changes send to LKML
https://lore.kernel.org/patchwork/patch/1012205/
https://lore.kernel.org/patchwork/patch/1012206/

**Comment by feckert** [comment]
I have update my PR with changes v4 send to LKML
https://lore.kernel.org/patchwork/patch/1012867/
https://lore.kernel.org/patchwork/patch/1012866/

**Comment by miczyg1** [comment]
@feckert I will prepare dumps ASAP. Would You mind sending me the correct openwrt config so i can easily build the image for apu2? I remember I haven't set all needed options last time and would like to not waste much time on it going back to topic.

**Comment by feckert** [comment]
@miczyg1 thanks for feedback but it is not easy for me because we have a very customized openwrt config with a lot of additional feeds.
I think it is easier for me and you to checkout a vanilla openwrt master and select the **target** x86 and **subtarget** 64 and then apply on top my pullrequest (cherrypick).

**Comment by miczyg1** [comment]
@feckert that is what I basically did last time. However later I have found out that the generic config for x86_x64 did not select few GPIO options and I had to build image like 5 times.

**Comment by feckert** [comment]
I will send you tomorrow at stripped openwrt config

Am Mi., 28. Nov. 2018, 15:15 hat Michał Żygowski <notifications@github.com>
geschrieben:

**Comment by ynezz** [comment]
@miczyg1  Adding following

```
CONFIG_TARGET_x86=y
CONFIG_TARGET_x86_64=y
CONFIG_TARGET_x86_64_Generic=y
CONFIG_PACKAGE_e100-firmware=y
CONFIG_PACKAGE_kmod-3c59x=y
CONFIG_PACKAGE_kmod-8139cp=y
CONFIG_PACKAGE_kmod-8139too=y
CONFIG_PACKAGE_kmod-e100=y
CONFIG_PACKAGE_kmod-forcedeth=y
CONFIG_PACKAGE_kmod-gpio-button-hotplug=y
CONFIG_PACKAGE_kmod-gpio-nct5104d=y
CONFIG_PACKAGE_kmod-leds-apu2=y
CONFIG_PACKAGE_kmod-leds-gpio=y
CONFIG_PACKAGE_kmod-libphy=y
CONFIG_PACKAGE_kmod-natsemi=y
CONFIG_PACKAGE_kmod-ne2k-pci=y
CONFIG_PACKAGE_kmod-pcnet32=y
CONFIG_PACKAGE_kmod-sis900=y
CONFIG_PACKAGE_kmod-sky2=y
CONFIG_PACKAGE_kmod-slhc=m
CONFIG_PACKAGE_kmod-sp5100_tco=y
CONFIG_PACKAGE_kmod-tg3=y
CONFIG_PACKAGE_kmod-tulip=y
CONFIG_PACKAGE_kmod-usb-core=y
CONFIG_PACKAGE_kmod-usb-ehci=y
CONFIG_PACKAGE_kmod-usb-ohci=y
CONFIG_PACKAGE_kmod-usb-serial=y
CONFIG_PACKAGE_kmod-usb2=y
CONFIG_PACKAGE_kmod-usb3=y
CONFIG_PACKAGE_kmod-via-rhine=y
CONFIG_PACKAGE_kmod-via-velocity=y
```

into `.config` and running `make defconfig` should produce working image for `apu2`.

**Comment by feckert** [comment]
@miczyg1 @pietrushnic Could someone please send Andy Shevchenko the ACPI dump from coreboot of the APU board to this mail https://www.spinics.net/lists/linux-gpio/msg34771.html ?
@pietrushnic you did already send this to https://pastebin.com/2neC1Sri but the file is not visible anymore.

**Comment by miczyg1** [comment]
@feckert is it required that the dump has to be done under Openwrt?

**Comment by feckert** [comment]
@miczyg1 no this is not necessary. He hints only the ACPI information for the APU-BIOS (coreboot)

**Comment by miczyg1** [comment]
@feckert I have sent the logs, but I guess I couldn't hook well into the thread despite setting the same in-reply-to... I wish this mailing list could have a reply link.

In case something gone wrong...
[apu2_acpidump](https://cloud.3mdeb.com/index.php/s/MxBZFxK9f85ZKb3/download)
[lspci.log](https://cloud.3mdeb.com/index.php/s/XmQ9J67wAsBXQ3B/download)

**Comment by feckert** [comment]
@miczyg1 thank you very much

**Comment by andy-shev** [comment]
Thanks for ACPI tables. Looks pretty basic. I sent to mailing list the proposed changes.
The platform driver for GPIO keys shouldn't be exist at all if ACPI tables were written correctly from the beginning.

**Comment by feckert** [comment]
@miczyg1 @pietrushnic I have a talk with @andy-shev last week on the linux-gpio mailinglist.
https://www.spinics.net/lists/linux-gpio/msg34887.html
Pleas have a look on @andy-shev suggestion to get the ACPI table fixed in APUx coreboot.

The driver for gpio handling is pending.
I'm not sure how we're gonna move on now.

**Comment by ynezz** [comment]
@feckert Any news? How are you going to proceed with this PR?

**Comment by feckert** [comment]
@ynezz thank you for asking

I haven't heard anything from upstream (linux) anymore and haven't asked any more.

The problem is that the ACPI-tables from the APU`s are wrong. The maintainer @andy-shev of the GPIO-subsystem wants to cover all BIOS version and feature version before merging this driver. The ACPI thing is missing!

The driver is ready and works with the current BIOS of PC-Engine with APU4 boards this is the one I have tested.

I don't know anything about ACPI and have not enough time. And I also don't get support from the coreboot people who makes the feature BIOS for the APU!

**Comment by andy-shev** [comment]
This is in upstream now.
https://github.com/dvhart/linux-pdx86/blob/for-next/drivers/platform/x86/pcengines-apuv2.c

AFAIU above the issue is to bring a new generation to be supported. Yes, I would like to see BIOS fixed for that, rather than continuation of hard coding stuff here and there.

**Comment by miczyg1** [comment]
@feckert we are sorry but we have limited resources for APU BIOS development. We will schedule the fixes for ACPI and hopefully address it in next month. 

ACPI tables are pretty basic because coreboot does only basic initialization of hardware and creates only necessary ACPI tables. GPIOs aren't that significant. We are facing much more complex issues which affect platform usability. We can at least provide the correct ACPI tables for SoC GPIOs in a reasonable amount of time. Hopefully, next month if we do not face any other higher priority items which require immediate action.

**Comment by feckert** [comment]
@andy-shev 
I haven't been following the whole thing upstream in linux lately.
That's what my driver would have done too.
Nevertheless, I have to admit that the new one who is merged upstream is more elegant :-)

Apart from that the mpcie2/mpcie3 reset pins are missing.
Maybe i will send a patch to add this.

If this is already merged upstream then we should remove the https://github.com/torvalds/linux/blob/master/drivers/leds/leds-apu.c driver partly, because APU2/APU3/APU4 is handled in the new driver. As I can see APU1 is not handled by the new driver.
https://github.com/dvhart/linux-pdx86/blob/for-next/drivers/platform/x86/pcengines-apuv2.c#L25-L26

@miczyg1 
> I don't know anything about ACPI and have not enough time. And I also don't get support from the coreboot people who makes the feature BIOS for the APU!

Sorry That was brashly formulated.

**Comment by feckert** [comment]
@ynezz as mention in my last reply I think we could close this pullrequest and maybe backport the new upstream APU gpio driver and platform definitions. We could also add the missing mpcie2/mpcie3 reset gpio pins as a patch to the new driver. This is the only thing what is missing and what my driver had in addition.

This new driver will come to openwrt if we update the linux kernel to 5.1 or greater.

**Comment by ynezz** [comment]
> We could also add the missing mpcie2/mpcie3 reset gpio pins as a patch to the new driver. 

Ok, but it would be nice to get this first accepted upstream, afterwards backport it to OpenWrt.

> I think we could close this pullrequest

Ok, thank you all!

**Comment by andy-shev** [comment]
> @andy-shev
> Apart from that the mpcie2/mpcie3 reset pins are missing.
> Maybe i will send a patch to add this.

Please, do!

> If this is already merged upstream then we should remove the https://github.com/torvalds/linux/blob/master/drivers/leds/leds-apu.c driver partly, because APU2/APU3/APU4 is handled in the new driver. As I can see APU1 is not handled by the new driver.
> https://github.com/dvhart/linux-pdx86/blob/for-next/drivers/platform/x86/pcengines-apuv2.c#L25-L26

Patches are welcome, you know!

**Pull request closed without merging**